aesd-char-driver/main.c

```diff
@@ -0,0 +1,269 @@
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+
+    // didn't do anything in open, don't need to do anything in release
+
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t bytes_read = 0;
+    struct aesd_dev *dev = (struct aesd_dev *)filp->private_data;
+    size_t offset = 0;
+    struct aesd_buffer_entry *entry;
+
+    if(buf == NULL || filp == NULL || f_pos == NULL){
+        PDEBUG("invalid pointer input to read\n");
+        return -EINVAL;
+    }
+
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+
+    if(count == 0){
+        return 0;
+    }
+
+    mutex_lock(&dev->mtx);
+
+    // find the line and offset for the desired pos
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->buf, *f_pos, &offset);
+
+    // If an entry isn't found, (offset too big) we've reached the end of the buffer
+    // and can't read anything
+    if(entry == NULL){
+        mutex_unlock(&dev->mtx);
+        return 0;
+    }
+
+    // only read count bytes if that's smaller than what's remaining on the 
+    // current line
+    bytes_read = min(count, entry->size - offset);
+
+    // actually copy the data out
+    if(copy_to_user(buf, entry->buffptr+offset, bytes_read)){
+        PDEBUG("failed to read data into user memory\n");
+        mutex_unlock(&dev->mtx);
+        return -EFAULT;
+    }
+
+    // adjust position
+    *f_pos += bytes_read;
+
+    mutex_unlock(&dev->mtx);
+    return bytes_read;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t bytes_written = 0;
+    const char *buf_to_free;
+    struct aesd_dev *dev = (struct aesd_dev *)filp->private_data;
+
+    if(buf == NULL || filp == NULL || f_pos == NULL){
+        PDEBUG("invalid pointer input to write\n");
+        return -EINVAL;
+    }
+
+    PDEBUG("write %zu bytes with offset %lld\n",count,*f_pos);
+
+    if(count == 0){
+        return 0;
+    }
+
+    // lock it down
+    mutex_lock(&dev->mtx);
+
+    // Allocate kernel memory for incoming data (expand or create line buf)
+    if(dev->line_len){
+        // line already started
+        dev->line_buf = krealloc(dev->line_buf, dev->line_len + count, GFP_KERNEL);
+        if(dev->line_buf == NULL){
+            PDEBUG("failed to widen line buf\n");
+            mutex_unlock(&dev->mtx);
+            return -ENOMEM;
+        }
+    }
+    else{
+        // brand new line
+        dev->line_buf = kmalloc(count, GFP_KERNEL);
+        if(dev->line_buf == NULL){
+            PDEBUG("failed to create line buf\n");
+            mutex_unlock(&dev->mtx);
+            return -ENOMEM;
+        }
+    }
+
+    // loop through write data and handle complete lines as they come
+    while(bytes_written < count){
+        if(get_user(dev->line_buf[dev->line_len++], &buf[bytes_written++])){
```

Comment:
@jbohn3353 I think you could simplify the logic here with `copy_from_user`/`strnchr`